### PR TITLE
IAP + CDN 

### PR DESCRIPTION
**Pull request by rramkumar1**
This is the last PR for IAP + CDN integration.

TODO's in separate PR's:
    1. Need to write an e2e test using new test framework.
    2. Need to add more unit tests to verify behavior.

---
This change is [Reviewable](https://reviewable.io/reviews/kubernetes/ingress-gce/301)

### CODE REVIEW

**Comment by ConradIrwin** [comment]
Thanks @rramkumar1 looking forward to this :D.

File: pkg/backends/features/cdn.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
Comment (MrHohn) [inline pkg/backends/features/cdn.go]: 2018

File: pkg/backends/features/cdn.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+// to the passed in compute.BackendService. A GCE API call still needs to be
+// made to actually persist the changes.
+func applyCDNSettings(sp utils.ServicePort, be *composite.BackendService) {
+	beConfig := sp.BackendConfig
+	setCDNDefaults(beConfig)
+	// Apply the boolean switch
+	be.EnableCDN = beConfig.Spec.Cdn.Enabled
+	// Apply the cache key policies
+	cacheKeyPolicy := beConfig.Spec.Cdn.CachePolicy
+	be.CdnPolicy = &composite.BackendServiceCdnPolicy{CacheKeyPolicy: &composite.CacheKeyPolicy{}}
```
Comment (MrHohn) [inline pkg/backends/features/cdn.go]: Curious is it valid to have `be.EnableCDN=false` but `be.CdnPolicy=non-nil`?

File: pkg/backends/features/cdn.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	return false
+}
+
+// applyCDNSettings applies the CDN settings specified in the BackendConfig
+// to the passed in compute.BackendService. A GCE API call still needs to be
+// made to actually persist the changes.
+func applyCDNSettings(sp utils.ServicePort, be *composite.BackendService) {
+	beConfig := sp.BackendConfig
+	setCDNDefaults(beConfig)
```
Comment (MrHohn) [inline pkg/backends/features/cdn.go]: So this is applying defaults to backendConfig but not backendService, I was confused :)

File: pkg/composite/composite.go [hunk 4]
```diff
@@ -240,6 +240,13 @@ func (be *BackendService) toAlpha() (*computealpha.BackendService, error) {
 	if err != nil {
 		return nil, fmt.Errorf("error unmarshalling BackendService JSON to compute alpha type: %v", err)
 	}
+	// Set force send fields. This is a temporary hack.
+	if alpha.CdnPolicy != nil && alpha.CdnPolicy.CacheKeyPolicy != nil {
+		alpha.CdnPolicy.CacheKeyPolicy.ForceSendFields = []string{"IncludeHost", "IncludeProtocol", "IncludeQueryString", "QueryStringBlaclist", "QueryStringWhitelist"}
```
Comment (MrHohn) [inline pkg/composite/composite.go]: QueryStringBlaclist -> QueryStringBlacklist, same for below.

File: pkg/backends/features/iap.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	beTemp := &composite.BackendService{}
+	applyIAPSettings(sp, beTemp)
+	// We need to compare the SHA256 of the client secret instead of the client secret itself
+	// since that field is redacted when getting a BackendService.
+	beTemp.Iap.Oauth2ClientSecretSha256 = fmt.Sprintf("%x", sha256.Sum256([]byte(beTemp.Iap.Oauth2ClientSecret)))
+	if be.Iap == nil || beTemp.Iap.Enabled != be.Iap.Enabled || beTemp.Iap.Oauth2ClientId != be.Iap.Oauth2ClientId || beTemp.Iap.Oauth2ClientSecretSha256 != be.Iap.Oauth2ClientSecretSha256 {
+		glog.V(3).Infof("BackendService is %+v", be)
+		applyIAPSettings(sp, be)
+		glog.V(3).Infof("Updated IAP settings for service %v/%v.", sp.ID.Service.Namespace, sp.ID.Service.Name)
```
Comment (MrHohn) [inline pkg/backends/features/iap.go]: Maybe V(2)? Seems worth to log it if an update is needed.

Also `Updated` might not be accurate? As the actual update will happen after this.

File: pkg/backends/features/iap.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	if sp.BackendConfig == nil {
+		return false
+	}
+	beTemp := &composite.BackendService{}
+	applyIAPSettings(sp, beTemp)
+	// We need to compare the SHA256 of the client secret instead of the client secret itself
+	// since that field is redacted when getting a BackendService.
+	beTemp.Iap.Oauth2ClientSecretSha256 = fmt.Sprintf("%x", sha256.Sum256([]byte(beTemp.Iap.Oauth2ClientSecret)))
+	if be.Iap == nil || beTemp.Iap.Enabled != be.Iap.Enabled || beTemp.Iap.Oauth2ClientId != be.Iap.Oauth2ClientId || beTemp.Iap.Oauth2ClientSecretSha256 != be.Iap.Oauth2ClientSecretSha256 {
+		glog.V(3).Infof("BackendService is %+v", be)
```
Comment (MrHohn) [inline pkg/backends/features/iap.go]: Is this necessary?

File: pkg/backends/features/iap.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	backendconfigv1beta1 "k8s.io/ingress-gce/pkg/apis/backendconfig/v1beta1"
+	"k8s.io/ingress-gce/pkg/composite"
+	"k8s.io/ingress-gce/pkg/utils"
+)
+
+// EnsureIAP reads the IAP configuration specified in the BackendConfig
+// and applies it to the BackendService if it is stale. It returns true
+// if there were existing settings on the BackendService that were overwritten.
+func EnsureIAP(sp utils.ServicePort, be *composite.BackendService) bool {
+	if sp.BackendConfig == nil {
```
Comment (MrHohn) [inline pkg/backends/features/iap.go]: I might be missing something...How is the disable case handled? Similar question for CDN as well.

File: pkg/backends/features/cdn.go (hunk 1 above)
Comment (rramkumar1) [reply pkg/backends/features/cdn.go]: Done

File: pkg/backends/features/cdn.go (hunk 3 above)
Comment (rramkumar1) [reply pkg/backends/features/cdn.go]: Yep, we apply defaults to the BackendConfig so that in the apply[IAP,CDN]Settings functions we don't have to have ugly nil checks.

File: pkg/backends/features/cdn.go (hunk 2 above)
Comment (rramkumar1) [reply pkg/backends/features/cdn.go]: Yep. If the user disabled CDN but provided cache policies, then we will act on it accordingly.

File: pkg/backends/features/iap.go (hunk 6 above)
Comment (rramkumar1) [reply pkg/backends/features/iap.go]: Oops, was debugging something and forgot to delete.

File: pkg/backends/features/iap.go (hunk 5 above)
Comment (rramkumar1) [reply pkg/backends/features/iap.go]: Done

File: pkg/composite/composite.go (hunk 4 above)
Comment (rramkumar1) [reply pkg/composite/composite.go]: Nice catch :)

File: pkg/backends/features/iap.go (hunk 7 above)
Comment (rramkumar1) [reply pkg/backends/features/iap.go]: Discussed this offline but we should probably document that user should remove the annotation from the service first?

File: pkg/backends/backends.go [hunk 8]
```diff
@@ -101,14 +103,16 @@ func NewBackendPool(
...
-		negGetter:     negGetter,
-		nodePool:      nodePool,
-		healthChecker: healthChecker,
-		namer:         namer,
+		cloud:                cloud,
+		negGetter:            negGetter,
+		nodePool:             nodePool,
+		healthChecker:        healthChecker,
+		namer:                namer,
+		backendConfigEnabled: backendConfigEnabled,
```
Comment (MrHohn) [inline pkg/backends/backends.go]: Might be worth to separate this out to another PR if we can't get this in soon..

**Comment by ConradIrwin** [comment]
w00w00! thanks all :D

**Comment by rramkumar1** [comment]
@ConradIrwin Note that this is not officially available yet. It will only be released when we cut a new ingress-gce release, which will be 1.2.0

**Comment by ConradIrwin** [comment]
Noted, thank you. Do you have an estimate on that timeline?

I'm just excited to see you're working through the backlog of feature requests :). (and I'd cheekily put another nudge in for #33, but don't block shipping this on implementing that :D)

**Comment by rramkumar1** [comment]
@ConradIrwin A very conservative estimate would be early July.
